Static checker fails for `ingress_per_unit`

Prometheus `static-charm` check fails for `ingress_per_unit`.

Meanwhile  I added the following to pyproject.toml
```toml
[[tool.mypy.overrides]]
module = ["charms.traefik_k8s.*"]
follow_imports = "silent"
```

Might be a good idea to add static checks here.

```
lib/charms/traefik_k8s/v0/ingress_per_unit.py:229: error: Type variable "charms.traefik_k8s.v0.ingress_per_unit._IngressPerUnitBase._IngressPerUnitEventType" is unbound  [valid-type]
        on: _IngressPerUnitEventType
            ^
lib/charms/traefik_k8s/v0/ingress_per_unit.py:229: note: (Hint: Use "Generic[_IngressPerUnitEventType]" or "Protocol[_IngressPerUnitEventType]" base class to bind "_IngressPerUnitEventType" inside a class)
lib/charms/traefik_k8s/v0/ingress_per_unit.py:229: note: (Hint: Use "_IngressPerUnitEventType" in function signature to bind "_IngressPerUnitEventType" inside a function)
lib/charms/traefik_k8s/v0/ingress_per_unit.py:263: error: _IngressPerUnitEventType? has no attribute "ready"  [attr-defined]
                self.on.ready.emit(relation)
                ^
lib/charms/traefik_k8s/v0/ingress_per_unit.py:265: error: _IngressPerUnitEventType? has no attribute "available"  [attr-defined]
                self.on.available.emit(relation)
                ^
lib/charms/traefik_k8s/v0/ingress_per_unit.py:267: error: _IngressPerUnitEventType? has no attribute "failed"  [attr-defined]
                self.on.failed.emit(relation)
                ^
lib/charms/traefik_k8s/v0/ingress_per_unit.py:288: error: _IngressPerUnitEventType? has no attribute "broken"  [attr-defined]
            self.on.broken.emit(event.relation)
            ^
lib/charms/traefik_k8s/v0/ingress_per_unit.py:301: error: Item "None" of "Optional[Any]" has no attribute "name"  [union-attr]
            if not relation.app.name:
                   ^
lib/charms/traefik_k8s/v0/ingress_per_unit.py:504: error: Incompatible return value type (got "Dict[Any, Dict[Any, Any]]", expected "Dict[Unit, RequirerData]")  [return-value]
            return requirer_unit_data
                   ^
lib/charms/traefik_k8s/v0/ingress_per_unit.py:604: error: Incompatible types in assignment (expression has type "None", variable has type "Tuple[Optional[str], int]")  [assignment]
                self._auto_data = None
                                  ^
lib/charms/traefik_k8s/v0/ingress_per_unit.py:667: error: Item "None" of "Optional[Relation]" has no attribute "data"  [union-attr]
                raw = self.relation.data[self.unit].get("data")
                      ^
lib/charms/traefik_k8s/v0/ingress_per_unit.py:710: error: Item "None" of "Optional[Relation]" has no attribute "data"  [union-attr]
            self.relation.data[self.unit]["data"] = yaml.safe_dump(data)
            ^
lib/charms/traefik_k8s/v0/ingress_per_unit.py:320: error: Unused "type: ignore" comment
            if not relation.app.name:  # type: ignore
            ^
Found 11 errors in 1 file (checked 2 source files)
```

--- a/.github/workflows/pull-request.yaml
+++ b/.github/workflows/pull-request.yaml
@@ -70,3 +70,6 @@ jobs:
         if: failure()
         with:
           app: traefik-k8s
+
+      - name: Static IPU checks
+        run: tox -e static-ipu

--- a/lib/charms/traefik_k8s/v0/ingress_per_unit.py
+++ b/lib/charms/traefik_k8s/v0/ingress_per_unit.py
@@ -47,7 +47,8 @@ class SomeCharm(CharmBase):
 ```
 """
 import logging
-from typing import Dict, Optional, TypeVar, Union
+import typing
+from typing import Dict, Optional, Union, Type
 
 import ops.model
 import yaml
@@ -133,7 +134,7 @@ except ImportError:
     from typing_extensions import TypedDict  # py35 compat
 
 
-class RequirerData(TypedDict):
+class RequirerData(TypedDict):  # pyright: reportGeneralTypeIssues=false
     """Model of the data a unit implementing the requirer will need to provide."""
 
     model: str
@@ -225,8 +226,9 @@ class IngressPerUnitEvents(ObjectEvents):
 class _IngressPerUnitBase(Object):
     """Base class for IngressPerUnit interface classes."""
 
-    _IngressPerUnitEventType = TypeVar("_IngressPerUnitEventType", bound=IngressPerUnitEvents)
-    on: _IngressPerUnitEventType
+    if typing.TYPE_CHECKING:
+        @property
+        def on(self) -> IngressPerUnitEvents: ...  # noqa
 
     def __init__(self, charm: CharmBase, relation_name: str = DEFAULT_RELATION_NAME):
         """Constructor for _IngressPerUnitBase.
@@ -290,14 +292,15 @@ class _IngressPerUnitBase(Object):
     def _handle_upgrade_or_leader(self, _):
         pass
 
-    def is_available(self, relation: Relation = None) -> bool:
+    def is_available(self, relation: Optional[Relation] = None) -> bool:
         """Check whether the given relation is available.
 
         Or any relation if not specified.
         """
         if relation is None:
             return any(map(self.is_available, self.relations))
-
+        if relation.app is None:
+            return False
         if not relation.app.name:
             # Juju doesn't provide JUJU_REMOTE_APP during relation-broken
             # hooks. See https://github.com/canonical/operator/issues/693.
@@ -306,24 +309,20 @@ class _IngressPerUnitBase(Object):
 
         return True
 
-    def is_ready(self, relation: Relation = None) -> bool:
+    def is_ready(self, relation: Optional[Relation] = None) -> bool:
         """Checks whether the given relation is ready.
 
         Or any relation if not specified.
         A given relation is ready if the remote side has sent valid data.
+        The base implementation does nothing but check that the relation is
+        available. It's up to subclasses to decide what it means for the
+        relation to be actually 'ready'.
         """
         if relation is None:
             return any(map(self.is_ready, self.relations))
-        if relation.app is None:
-            # No idea why, but this happened once.
-            return False
-        if not relation.app.name:  # type: ignore
-            # Juju doesn't provide JUJU_REMOTE_APP during relation-broken
-            # hooks. See https://github.com/canonical/operator/issues/693
-            return False
-        return True
+        return self.is_available(relation)
 
-    def is_failed(self, _: Relation = None) -> bool:
+    def is_failed(self, _: Optional[Relation] = None) -> bool:
         """Checks whether the given relation is failed.
 
         Or any relation if not specified.
@@ -361,7 +360,7 @@ class IngressPerUnitProvider(_IngressPerUnitBase):
             log.info("shared supported_versions shim information")
             relation.data[self.charm.app]["_supported_versions"] = "- v1"
 
-    def is_ready(self, relation: Relation = None) -> bool:
+    def is_ready(self, relation: Optional[Relation] = None) -> bool:
         """Checks whether the given relation is ready.
 
         Or any relation if not specified.
@@ -381,7 +380,7 @@ class IngressPerUnitProvider(_IngressPerUnitBase):
 
         return any(requirer_unit_data.values())
 
-    def is_failed(self, relation: Relation = None) -> bool:
+    def is_failed(self, relation: Optional[Relation] = None) -> bool:
         """Checks whether the given relation is failed.
 
         Or any relation if not specified.
@@ -630,9 +629,11 @@ class IngressPerUnitRequirer(_IngressPerUnitBase):
     @property
     def relation(self) -> Optional[Relation]:
         """The established Relation instance, or None if still unrelated."""
+        if len(self.relations) > 1:
+            raise ValueError('Multiple ingress-per-unit relations found.')
         return self.relations[0] if self.relations else None
 
-    def is_ready(self, relation: Relation = None) -> bool:
+    def is_ready(self, relation: Optional[Relation] = None) -> bool:
         """Checks whether the given relation is ready.
 
         Or any relation if not specified.
@@ -643,7 +644,7 @@ class IngressPerUnitRequirer(_IngressPerUnitBase):
 
         return bool(self.url)
 
-    def is_failed(self, relation: Relation = None) -> bool:
+    def is_failed(self, relation: Optional[Relation] = None) -> bool:
         """Checks whether the given relation is failed.
 
         Or any relation if not specified.
@@ -664,7 +665,7 @@ class IngressPerUnitRequirer(_IngressPerUnitBase):
 
         try:
             # grab the data and validate it; might raise
-            raw = self.relation.data[self.unit].get("data")
+            raw = relation.data[self.unit].get("data")
         except Exception:
             log.exception("Error accessing relation databag")
             return True
@@ -707,6 +708,9 @@ class IngressPerUnitRequirer(_IngressPerUnitBase):
             "port": port,
         }
         _validate_data(data, INGRESS_REQUIRES_UNIT_SCHEMA)
+
+        if not self.relation:
+            raise RuntimeError("Can't publish ingress data: no relation found.")
         self.relation.data[self.unit]["data"] = yaml.safe_dump(data)
 
     @property

--- a/lib/charms/traefik_k8s/v0/ingress_per_unit.py
+++ b/lib/charms/traefik_k8s/v0/ingress_per_unit.py
@@ -48,7 +48,7 @@ class SomeCharm(CharmBase):
 """
 import logging
 import typing
-from typing import Dict, Optional, Union, Type
+from typing import Dict, Optional, Union
 
 import ops.model
 import yaml
@@ -227,8 +227,10 @@ class _IngressPerUnitBase(Object):
     """Base class for IngressPerUnit interface classes."""
 
     if typing.TYPE_CHECKING:
+
         @property
-        def on(self) -> IngressPerUnitEvents: ...  # noqa
+        def on(self) -> IngressPerUnitEvents:
+            ...  # noqa
 
     def __init__(self, charm: CharmBase, relation_name: str = DEFAULT_RELATION_NAME):
         """Constructor for _IngressPerUnitBase.
@@ -630,7 +632,7 @@ class IngressPerUnitRequirer(_IngressPerUnitBase):
     def relation(self) -> Optional[Relation]:
         """The established Relation instance, or None if still unrelated."""
         if len(self.relations) > 1:
-            raise ValueError('Multiple ingress-per-unit relations found.')
+            raise ValueError("Multiple ingress-per-unit relations found.")
         return self.relations[0] if self.relations else None
 
     def is_ready(self, relation: Optional[Relation] = None) -> bool:

--- a/tox.ini
+++ b/tox.ini
@@ -76,3 +76,10 @@ deps =
     -r{toxinidir}/requirements.txt
 commands =
     pytest -v --tb native --ignore={[vars]tst_path}unit --log-cli-level=INFO -s {posargs}
+
+[testenv:static-ipu]
+description = Static code checking for ingress-per-unit
+deps =
+    pyright
+    -r{toxinidir}/requirements.txt
+commands = pyright {toxinidir}/lib/charms/traefik_k8s/v0/ingress_per_unit.py

